### PR TITLE
Refactor mail notification

### PR DESCRIPTION
**Pull request by lunny**
extract from #4001

### CODE REVIEW

**Comment by lunny** [comment]
extract from #4001

**Review by zeripath — COMMENTED** [review]
I don't much like having the multiple empty functions. I guess we need to move to more generic notify interface with typed event objects? Does accepting this PR help with that move or Is there a current bug that needs this?

I guess this PR helps to place things in to standard structure but I think that this might be better shown as a commit in a well structured PR that would compile and pass the tests at each step along the way. Was this extracted from such a larger commit?

File: modules/notification/mail/mail.go [hunk 1]
```diff
@@ -0,0 +1,126 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
```
Comment (kolaente) [inline modules/notification/mail/mail.go]: Could you update the date to 2019?

**Comment by zeripath** [comment]
OK, can we create a `NullNotifier` that has these empty methods, then change the definition of `mailNotifier` to be:

```
type mailNotifier {
   NullNotifier
}
```

and then just reimplement the methods that need reimplementing? I appreciate that that just moves these empty methods elsewhere but then implementing a notifier becomes much easier.

**Comment by lunny** [comment]
@zeripath good idea & done.

**Review by zeripath — APPROVED** [review]
Thanks @lunny !

**Review by lafriks — APPROVED** [review]

**Pull request merged**